# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/nsDSURIContentListener.cpp
# Commit: baffa8a7b961
# Full Hash: baffa8a7b9618e9a02e4bd808adb71dead99cce5
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2019-11-12 09:43:07
# Regressor Bug: 1589270
# File Overlap Count: 1
# Description:
#   Bug 1589270 - Part 3: Convert nsExternalHelperApp to use BrowsingContext instead of nsIInterfaceRequestor. r=bzbarsky
#   
#   This also converts MaybeCloseWindowHelper, and results in the window close operations being always run in the parent (even without DocumentChannel).
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D49528
# ==============================================================================

diff -r 68bc2ae9b89a -r baffa8a7b961 docshell/base/nsDSURIContentListener.cpp
--- a/docshell/base/nsDSURIContentListener.cpp	Fri Nov 08 03:01:03 2019 +0000
+++ b/docshell/base/nsDSURIContentListener.cpp	Fri Nov 08 04:35:05 2019 +0000
@@ -20,6 +20,7 @@
 #include "nsIMultiPartChannel.h"
 
 using namespace mozilla;
+using namespace mozilla::dom;
 
 NS_IMPL_ADDREF(MaybeCloseWindowHelper)
 NS_IMPL_RELEASE(MaybeCloseWindowHelper)
@@ -28,10 +29,8 @@
   NS_INTERFACE_MAP_ENTRY(nsISupports)
 NS_INTERFACE_MAP_END
 
-MaybeCloseWindowHelper::MaybeCloseWindowHelper(
-    nsIInterfaceRequestor* aContentContext)
-    : mContentContext(aContentContext),
-      mWindowToClose(nullptr),
+MaybeCloseWindowHelper::MaybeCloseWindowHelper(BrowsingContext* aContentContext)
+    : mBrowsingContext(aContentContext),
       mTimer(nullptr),
       mShouldCloseWindow(false) {}
 
@@ -41,36 +40,32 @@
   mShouldCloseWindow = aShouldCloseWindow;
 }
 
-nsIInterfaceRequestor* MaybeCloseWindowHelper::MaybeCloseWindow() {
-  nsCOMPtr<nsPIDOMWindowOuter> window = do_GetInterface(mContentContext);
-  NS_ENSURE_TRUE(window, mContentContext);
-
+BrowsingContext* MaybeCloseWindowHelper::MaybeCloseWindow() {
   if (mShouldCloseWindow) {
     // Reset the window context to the opener window so that the dependent
     // dialogs have a parent
-    nsCOMPtr<nsPIDOMWindowOuter> opener =
-        nsGlobalWindowOuter::Cast(window)->GetSameProcessOpener();
+    RefPtr<BrowsingContext> opener = mBrowsingContext->GetOpener();
 
-    if (opener && !opener->Closed()) {
-      mContentContext = do_QueryInterface(opener);
+    if (opener && !opener->IsDiscarded()) {
+      mBCToClose = mBrowsingContext;
+      mBrowsingContext = opener;
 
       // Now close the old window.  Do it on a timer so that we don't run
       // into issues trying to close the window before it has fully opened.
       NS_ASSERTION(!mTimer, "mTimer was already initialized once!");
       NS_NewTimerWithCallback(getter_AddRefs(mTimer), this, 0,
                               nsITimer::TYPE_ONE_SHOT);
-      mWindowToClose = window;
     }
   }
-  return mContentContext;
+  return mBrowsingContext;
 }
 
 NS_IMETHODIMP
 MaybeCloseWindowHelper::Notify(nsITimer* timer) {
-  NS_ASSERTION(mWindowToClose, "No window to close after timer fired");
+  NS_ASSERTION(mBCToClose, "No window to close after timer fired");
 
-  mWindowToClose->Close();
-  mWindowToClose = nullptr;
+  mBCToClose->Close(CallerType::System, IgnoreErrors());
+  mBCToClose = nullptr;
   mTimer = nullptr;
 
   return NS_OK;
@@ -143,15 +138,11 @@
       aRequest->Cancel(NS_ERROR_DOM_BAD_URI);
       *aAbortProcess = true;
       // close the window since the navigation to a data URI was blocked
-      if (mDocShell) {
-        nsCOMPtr<nsIInterfaceRequestor> contentContext =
-            do_QueryInterface(mDocShell->GetWindow());
-        if (contentContext) {
-          RefPtr<MaybeCloseWindowHelper> maybeCloseWindowHelper =
-              new MaybeCloseWindowHelper(contentContext);
-          maybeCloseWindowHelper->SetShouldCloseWindow(true);
-          maybeCloseWindowHelper->MaybeCloseWindow();
-        }
+      if (mDocShell && mDocShell->GetBrowsingContext()) {
+        RefPtr<MaybeCloseWindowHelper> maybeCloseWindowHelper =
+            new MaybeCloseWindowHelper(mDocShell->GetBrowsingContext());
+        maybeCloseWindowHelper->SetShouldCloseWindow(true);
+        maybeCloseWindowHelper->MaybeCloseWindow();
       }
       return NS_OK;
     }